### PR TITLE
i#2207: enable Android cross build on Travis

### LINKED ISSUE #2207: add Android cross-compilation to Travis
Xref #1962

By running it in parallel, even if we have to download the NDK every time it still shouldn't increase the total Travis time.

### PR DESCRIPTION
**Pull request by fhahn**
For the Android cross build, we fetch Android NDK r10e and set it up to
use ld.bfd as ld for the build to work. Newer versions of the Android
NDK won't work until #1927 is fixed.

This commit introduces a new DYNAMORIO_CROSS_ANDROID_ONLY environment
variable, to allow building for Android on a separate instance, because
fetching the NDK might take a while.

Fixes #2207

### CODE REVIEW

**Comment by fhahn** [comment]
@derekbruening Travis is happy now I think! What do you think of using DYNAMORIO_CROSS_ANDROID_ONLY to build Android on a separate Travis instance? Setting up the NDK doesn't take too long, so we could be fine with adding it back to CROSS_ONLY. But we may want build different ABIs on travis in the future.

**Comment by derekbruening** [comment]
> @derekbruening Travis is happy now I think! What do you think of using DYNAMORIO_CROSS_ANDROID_ONLY to build Android on a separate Travis instance? Setting up the NDK doesn't take too long, so we could be fine with adding it back to CROSS_ONLY. But we may want build different ABIs on travis in the future.

The Android build seems to take 7 mins, and cross-arm is 10, so let's keep them separate (the goal is nothing over ~10 mins).

File: suite/runsuite.cmake [hunk 1]
```diff
@@ -314,33 +318,38 @@ if (UNIX AND ARCH_IS_X86)
...
+  endif (NOT cross_android_only)
   set(run_tests ${prev_run_tests})
 
   # Android cross-compilation and running of tests using "adb shell"
-  # FIXME i#2207: once we have Android cross-compilation working on Travis, remove this:
-  set(optional_cross_compile ON)
+  # For Travis cross_android_only builds, we want to fail on config failures.
+  # For user suite runs, we want to just skip if there's no cross setup.
+  if (NOT cross_android_only)
+      set(optional_cross_compile ON)
```
Comment (derekbruening) [inline suite/runsuite.cmake]: I think there's a bug here: optional_cross_compile is never reset to OFF after it's set to ON for NOT cross_only above.

File: suite/runsuite.cmake [hunk 2]
```diff
@@ -365,6 +374,15 @@ if (UNIX AND ARCH_IS_X86)
     set(android_extra_rel "")
     set(run_tests OFF) # build tests but don't run them
   endif ()
+
+  # Pass through toolchain file.
+  if(DEFINED ENV{DYNAMORIO_ANDROID_TOOLCHAIN})
```
Comment (derekbruening) [inline suite/runsuite.cmake]: style: if (

**Comment by fhahn** [comment]
> The Android build seems to take 7 mins, and cross-arm is 10, so let's keep them separate (the goal is nothing over ~10 mins).

Ok! Should I rename `DYNAMORIO_CROSS_ONLY` to something more fitting, e.g. `DYNAMORIO_CROSS_ARCHXX_LINUX_ONLY` ?

File: suite/runsuite.cmake (hunk 2 above)
Comment (fhahn) [inline suite/runsuite.cmake]: Done

File: suite/runsuite.cmake (hunk 1 above)
Comment (fhahn) [inline suite/runsuite.cmake]: Done, I've introduced a new variable storing the state of optional_cross_compile and use that to restore it later

**Comment by derekbruening** [comment]
> Should I rename DYNAMORIO_CROSS_ONLY to something more fitting, e.g. DYNAMORIO_CROSS_ARCHXX_LINUX_ONLY ?

Sure, that sounds good.

**Comment by fhahn** [comment]
I've renamed DYNAMORIO_CROSS_ONLY to DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY. It would be great if you could have a quick look at the last commit.

File: suite/runsuite.cmake [hunk 3]
```diff
@@ -319,30 +319,28 @@ if (UNIX AND ARCH_IS_X86)
   set(ENV{CXXFLAGS} "")
   set(prev_run_tests ${run_tests})
   set(run_tests OFF) # build tests but don't run them
-  if (NOT cross_android_only)
```
Comment (derekbruening) [inline suite/runsuite.cmake]: OK you're removing this b/c the "optional_cross_compile" will ignore the expected error b/c the Android Travis setup has no arm compiler

File: .travis.yml [hunk 4]
```diff
@@ -61,12 +61,12 @@ compiler:
 
 env:
   # With or without cross-compiling:
-  - DYNAMORIO_CROSS_ONLY=yes
+  - DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY=yes
```
Comment (derekbruening) [inline .travis.yml]: Not sure LINUX is necessary to be in there?  What is your logic there?

File: suite/runsuite.cmake (hunk 3 above)
Comment (fhahn) [inline suite/runsuite.cmake]: yep, so now the behavior for both cross-only options is the same, i.e. adding the configurations, but only failing if the respective variable is set.

File: .travis.yml (hunk 4 above)
Comment (fhahn) [inline .travis.yml]: Android builds are for ARM too, so I thought it would be a good idea to have Linux in there, to differentiate.

Comment (derekbruening) [inline .travis.yml]: Oh, right.  Sure.

**Pull request merged**